src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java

```diff
@@ -269,108 +279,189 @@ public StyleAccessLimits getAccessLimits(Authentication user, StyleInfo style) {
-        LOGGER.log(
-                Level.FINE,
-                "Returning {0} for layer {1} and user {2}",
-                new Object[] {limits, resource.prefixedName(), username});
+    // get the AccessInfo from the layer's containers returning a tuple with
+    // the Geometry of the allowed area if found and the less restrictive accessInfo
+    // among the ones of associated layerGroups
+    private Pair<Geometry, AccessInfo> getContainerRuleForLayerDirectAccess(
+            Object resource, Authentication user) {
+        MutablePair<Geometry, AccessInfo> result = new MutablePair<>(null, null);
+        Collection<LayerGroupContainmentCache.LayerGroupSummary> summaries;
+        if (resource instanceof ResourceInfo)
+            summaries = groupsCache.getContainerGroupsFor((ResourceInfo) resource);
+        else if (resource instanceof LayerInfo)
+            summaries = groupsCache.getContainerGroupsFor(((LayerInfo) resource).getResource());
+        else summaries = groupsCache.getContainerGroupsFor((LayerGroupInfo) resource);
+
+        for (LayerGroupContainmentCache.LayerGroupSummary gs : summaries) {
+            LayerGroupInfo.Mode mode = gs.getMode();
+            if (mode.equals(LayerGroupInfo.Mode.OPAQUE_CONTAINER)) {
+                // opaque mode deny access for the layer
+                AccessInfo newInfo = AccessInfo.DENY_ALL;
+                result.setValue(getLessRestrictiveRule(result.getRight(), newInfo));
+            } else if (!mode.equals(LayerGroupInfo.Mode.SINGLE)) {
+                // not opaque and not single mode, the container rule
+                // should override the layer rule
+                String workspace = gs.getWorkspace();
+                String layer = gs.getName();
+                RuleFilter filter = buildRuleFilter(workspace, layer, user);
+                AccessInfo newInfo = rules.getAccessInfo(filter);
+                newInfo = getLessRestrictiveRule(result.getRight(), newInfo);
+                result.setRight(newInfo);
+                if (newInfo.getAreaWkt() != null) {
+                    LayerGroupInfo gi = catalog.getLayerGroupByName(layer);
+                    CoordinateReferenceSystem crs = gi.getBounds().getCoordinateReferenceSystem();
+                    Geometry allowedArea = getReprojectArea(newInfo, crs);
+                    result.setLeft(allowedArea);
+                }
+            }
+        }
+        return result;
+    }
 
-        return limits;
+    // compares two access info and return the less restrictive one
+    private AccessInfo getLessRestrictiveRule(AccessInfo current, AccessInfo newInfo) {
+        if (newInfo == null) return current;
+        else if (current == null) return newInfo;
+        else if (current.getGrant().equals(GrantType.DENY)) return newInfo;
+        else if (current.getGrant().equals(GrantType.LIMIT)
+                && newInfo.getGrant().equals(GrantType.ALLOW)) return newInfo;
+        else return current;
```

Comment:
So if both are of type "limit", the less restrictive is the "current" one? Shouldn'it it somehow merge the rules, I don't know, pick the OR of the filters, the union of the geometries, the union of the list of attributes (and also say something about attribute read/write restrictions?). 

I'm not saying it's the thing to do, just wondering really. The way to go is to first clarify in the documentation what is the indeded behavior, and then implement it consistently.

Reply:
Also, if things are implemented consistently, I'm guessing you won't need pairs to bring aroud the AccessInfo (which contains a geometry itself) and the geometry, just build a new access info applying the aggregation rules.

Reply:
Thanks for your point @aaime. The LayerGroup doesn't have the possibility to define read write filters, so the thing to handle in that piece of is the allowedArea filter. 

I added the necessary logic to union the two geometries if the rules have the same grantType and have an allowedArea. If one of the container which might affect the layer (eg. NAMED) doesn't have an allowed area then no allowed area will be specified (the less restrictive).

Regarding the Pairs usage: the allowed area needs to be reprojected according to the LayerGroup CRS and inside the rule is served as WKT. The only way I found to handle this situation and pass to the Layer Limit building process a Geometry ready to be included in the filter was to use Pairs.